### PR TITLE
PR to pass solver with sklearn API

### PR DESCRIPTION
**Pull request by QB3**

### CODE REVIEW

File: examples/plot_grad_search_cv.py [hunk 1]
```diff
@@ -72,15 +73,19 @@
...
 print('sparse-ho started')
 
 t0 = time.time()
 Model = Lasso
 Criterion = CV
 log_alpha0 = np.log(alpha_max / 10)
 monitor_grad = Monitor()
-criterion = CrossVal(X, y, Lasso, cv=kf)
-algo = ImplicitForward(criterion, use_sk=True, max_iter=max_iter)
+criterion = CrossVal(X, y, Lasso, cv=kf, clf=clf)
```
Comment (agramfort) [inline examples/plot_grad_search_cv.py:87]: clf stands for classifier. Call it estimator

Comment (agramfort) [reply examples/plot_grad_search_cv.py:87]:
```suggestion
criterion = CrossVal(X, y, Model, cv=kf, clf=clf)
```
?

Comment (QB3) [reply examples/plot_grad_search_cv.py]: done thx

File: examples/plot_held_out_enet.py [hunk 2]
```diff
@@ -80,12 +80,12 @@
     print("lambda %i / %i" % (i, n_grid))
     for j in range(n_grid):
         print("lambda %i / %i" % (j, n_grid))
-        clf = ElasticNet_sk(
+        estimator = ElasticNet_sk(
```
Comment (agramfort) [inline examples/plot_held_out_enet.py:83]:
```
from sklearn import linear_model
estimator = linear_model.ElasticNet(...)
```
avoid weird names

File: examples/plot_held_out_lasso.py [hunk 3]
```diff
@@ -17,6 +17,7 @@
 import numpy as np
 import matplotlib.pyplot as plt
 import seaborn as sns
+import sklearn
```
Comment (agramfort) [inline examples/plot_held_out_lasso.py:20]: `from sklearn import linear_model`

File: sparse_ho/criterion.py [hunk 4]
```diff
@@ -463,12 +463,14 @@ def __init__(self, X, y, Model, cv=None, max_iter=1000):
             For int/None inputs, KFold is used.
         max_iter: int
             Maximal number of iteration for the state-of-the-art solver
+        estimator: sklearn Base Estimator
```
Comment (agramfort) [inline sparse_ho/criterion.py:466]:
```suggestion
        estimator: instance of ``sklearn.base.BaseEstimator``
```

File: sparse_ho/models.py [hunk 5]
```diff
@@ -21,22 +21,17 @@ class Lasso():
         Data.
     y: {ndarray, sparse matrix} of (n_samples)
         Target
-    TODO: other parameters should be remove
+    estimator: sklearn Base Estimator
```
Comment (agramfort) [inline sparse_ho/models.py:24]: instance ...

File: sparse_ho/models.py [hunk 6]
```diff
@@ -226,16 +221,17 @@ def get_L(X, is_sparse=False):
...
-                fit_intercept=False, max_iter=self.max_iter, warm_start=True)
-        self.clf.alpha = alpha
-        self.clf.tol = tol
-        # clf = linear_model.Lasso(
+        if self.estimator is None:
+            raise ValueError("You did not pass a solver with sklearn API")
+            # self.estimator = linear_model.Lasso(
+            #     fit_intercept=False, max_iter=self.max_iter, warm_start=True)
+        self.estimator.alpha = alpha
+        self.estimator.tol = tol
```
Comment (agramfort) [inline sparse_ho/models.py:229]: use `self.estimator.set_params(tol=tol, alpha=alpha)`

File: sparse_ho/models.py [hunk 7]
```diff
@@ -284,16 +280,17 @@ class wLasso():
         Data.
     y: {ndarray, sparse matrix} of (n_samples)
         Target
-    TODO: other parameters should be remove
+    estimator: sklearn Base Estimator
```
Comment (agramfort) [inline sparse_ho/models.py:283]: idem

File: sparse_ho/tests/test_grad_search_CV.py [hunk 8]
```diff
@@ -57,15 +57,16 @@
...
 
 
-def test_grad_search():
-    monitor = Monitor()
-    grad_search_CV(
-        X, y, Lasso, CV, ImplicitForward, log_alpha, monitor, n_outer=15)
+# def test_grad_search():
+#     monitor = Monitor()
+#     grad_search_CV(
+#         X, y, Lasso, CV, ImplicitForward, log_alpha, monitor, n_outer=15)
```
Comment (agramfort) [inline sparse_ho/tests/test_grad_search_CV.py:69]: clean up?

File: sparse_ho/tests/test_grad_search_CV.py [hunk 9]
```diff
@@ -96,5 +99,5 @@ def test_cross_val_criterion():
 
 
 if __name__ == '__main__':
-    test_grad_search()
+    # test_grad_search()
```
Comment (agramfort) [inline sparse_ho/tests/test_grad_search_CV.py:102]: ?

File: sparse_ho/models.py [hunk 10]
```diff
@@ -226,16 +221,17 @@ def get_L(X, is_sparse=False):
             return norm(X, axis=0) ** 2 / (X.shape[0])
 
     def sk(self, X, y, alpha, tol, max_iter):
```
Comment (agramfort) [inline sparse_ho/models.py:223]: this should not be a public method and should probably not be called sk. Next is to use celer here.

**Comment by QB3** [comment]
LGTM, wdyt @agramfort ?

File: sparse_ho/grad_search_CV.py [hunk 11]
```diff
@@ -10,7 +10,8 @@ def grad_search_CV(
         verbose=True, cv=5, random_state=0, test_size=0.33,
         tolerance_decrease='constant', tol=1e-5,
         t_max=10000, beta_star=None,):
-
+    """Depreciated fucntion, CV is now a proper criterion
```
Comment (agramfort) [inline sparse_ho/grad_search_CV.py:13]:
```suggestion
    """Deprecated function, CV is now a proper criterion
```
if it's deprecated should you just remove it?

File: sparse_ho/tests/test_elastic.py [hunk 12]
```diff
@@ -41,7 +41,13 @@
 alpha_2 = 0.01
 log_alpha1 = np.log(alpha_1)
 log_alpha2 = np.log(alpha_2)
-model = Elastic(X_train, y_train, log_alpha1, log_alpha2, max_iter=max_iter, tol=tol)
+
+model = Elastic(X_train, y_train, max_iter=max_iter, estimator=None)
```
Comment (agramfort) [inline sparse_ho/tests/test_elastic.py:45]: I find weird to call this Elastic and not ElasticNet

File: sparse_ho/forward.py [hunk 13]
```diff
@@ -9,13 +9,10 @@ class Forward():
     ----------
     criterion: criterion object
         HeldOIut, CrossVal or SURE
```
Comment (agramfort) [inline sparse_ho/forward.py:11]: HeldOIut -> Held-Out ?

File: expes/expe_fig_cross_val/main.py [hunk 14]
```diff
@@ -44,7 +44,7 @@
 # grad_search
 model = Lasso(X_train, y_train, np.log(alpha_max/10))
 criterion = CV(X_val, y_val, model, X_test=X_test, y_test=y_test)
-algo = ImplicitForward(criterion, use_sk=True)
```
Comment (agramfort) [inline expes/expe_fig_cross_val/main.py:47]: before you were using sklearn but now not as you do not pass estimator to model.

on purpose?

**Comment by agramfort** [comment]
thx @QB3 !

**Pull request merged**